manual/src/tutorials/parallelism.etex

```diff
@@ -0,0 +1,695 @@
+
+let sender () = Atomic.set r (Some "Hello")
+
+let rec receiver () =
+  match Atomic.get r with
+  | None -> Domain.cpu_relax (); receiver ()
+  | Some m -> print_endline m
+
+let main () =
+  let s = Domain.spawn sender in
+  let d = Domain.spawn receiver in
+  Domain.join s;
+  Domain.join d
+
+let _ = main ()
+\end{caml_example}
+
+While the sender and the receiver compete to access "r", this is not a data
+race since "r" is an atomic reference.
+
+\subsection{s:par_lockfree_stack}{Lock-free stack}
+
+The Atomic module is used to implement non-blocking, lock-free data structures.
+The following program implements a lock-free stack.
+
+\begin{caml_example*}{verbatim}
+module Lockfree_stack : sig
+  type 'a t
+  val make : unit -> 'a t
+  val push : 'a t -> 'a -> unit
+  val pop  : 'a t -> 'a option
+end = struct
+  type 'a t = 'a list Atomic.t
+
+  let make () = Atomic.make []
+
+  let rec push r v =
+    let s = Atomic.get r in
+    if Atomic.compare_and_set r s (v::s) then ()
+    else (Domain.cpu_relax (); push r v)
+
+  let rec pop r =
+    let s = Atomic.get r in
+    match s with
+    | [] -> None
+    | x::xs ->
+        if Atomic.compare_and_set r s xs then Some x
+        else (Domain.cpu_relax (); pop r)
+end
+\end{caml_example*}
+
+The atomic stack is represented by an atomic reference that holds a list. The
+"push" and "pop" operations use the "compare_and_set" primitive to attempt to
+atomically update the atomic reference. The expression "compare_and_set r seen
+v" sets the value of "r" to "v" if and only if its current value is physically
+equal to "seen". Importantly, the comparison and the update occur atomically.
+The expression evaluates to "true" if the comparison succeeded (and the update
+happened) and "false" otherwise.
+
+If the "compare_and_set" fails, then some other domain is also attempting to
+update the atomic reference at the same time. In this case, the "push" and
+"pop" operations call "Domain.cpu_relax" to back off for a short duration
+allowing competing domains to make progress before retrying the failed
+operation. This lock-free stack implementation is also known as Treiber
+stack.
+
+\section{s:par_sync}{Blocking synchronisation}
+
+Domains may also perform blocking synchronisation with the help of
+\stdmoduleref{Mutex}, \stdmoduleref{Condition} and \stdmoduleref{Semaphore}
+modules. These modules are the same as those used to synchronise threads
+created by the threads library~\ref{c:threads}. For clarity, in the rest of
+this chapter, we shall call the threads created by the threads library as
+systhreads. The following program implements a blocking version of a concurrent
+stack using mutex and condition variables.
+
+\begin{caml_example*}{verbatim}
+module Blocking_stack : sig
+  type 'a t
+  val make : unit -> 'a t
+  val push : 'a t -> 'a -> unit
+  val pop  : 'a t -> 'a
+end = struct
+  type 'a t = {
+    mutable contents: 'a list;
+    mutex : Mutex.t;
+    condition : Condition.t
```

Comment:
What about a more descriptive name for this condition variable?
```suggestion
    nonempty : Condition.t
```
Typically, there's only one mutex but several condition variables associated with a data structure, so different, descriptive names for the condvars are needed.